Gems/PhysX/Code/Source/EditorJointOptimizer.cpp

```diff
@@ -0,0 +1,166 @@
+        fittedLimit.m_childLocalRotation = childLocalRotation;
+        // value slightly less than pi to ensure limits are definitely inside allowed ranges
+        const float limitMax = AZ::Constants::Pi - 0.01f;
+        fittedLimit.m_swingLimitY = AZ::RadToDeg(AZ::GetClamp(swingLimitY, 0.0f, limitMax));
+        fittedLimit.m_swingLimitZ = AZ::RadToDeg(AZ::GetClamp(swingLimitZ, 0.0f, limitMax));
+
+        // twist values
+        float twistMin = AZ::Constants::Pi;
+        float twistMax = -AZ::Constants::Pi;
+        NumericalMethods::DoublePrecisionMath::Quaternion parentLocalConjugate = parentLocalRotation.GetConjugate();
+        for (const auto& localRotationSample : m_localRotationSamples)
+        {
+            NumericalMethods::DoublePrecisionMath::Quaternion relativeRotation = parentLocalConjugate * localRotationSample;
+            if (relativeRotation.GetZ() < 0.0)
+            {
+                relativeRotation = -relativeRotation;
+            }
+            float twist = aznumeric_cast<float>(2.0 * atan2(relativeRotation.GetX(), relativeRotation.GetW()));
+            twist = AZ::Wrap(twist, -AZ::Constants::Pi, AZ::Constants::Pi);
+            twistMin = AZ::GetMin(twistMin, twist);
+            twistMax = AZ::GetMax(twistMax, twist);
+        }
+        fittedLimit.m_twistLimitLower = AZ::RadToDeg(AZ::GetClamp(twistMin, -limitMax, limitMax));
+        fittedLimit.m_twistLimitUpper = AZ::RadToDeg(AZ::GetClamp(twistMax, -limitMax, limitMax));
+
+        return fittedLimit;
+    }
+
+    AZ::Outcome<double, NumericalMethods::Optimization::FunctionOutcome> D6JointLimitFitter::GetObjective(
+        const AZStd::vector<double>& x, bool debug) const
+    {
+        const auto parentLocalConjugate =
+            NumericalMethods::DoublePrecisionMath::Quaternion(x[0], x[1], x[2], x[3]).GetNormalized().GetConjugate();
+        const double swingLimitY = fabs(x[4]);
+        const double swingLimitZ = fabs(x[5]);
+        const double clampedLimitY = AZ::GetClamp(swingLimitY, 0.0, aznumeric_cast<double>(AZ::Constants::Pi));
+        const double clampedLimitZ = AZ::GetClamp(swingLimitZ, 0.0, aznumeric_cast<double>(AZ::Constants::Pi));
+        const double tanQuarterSwingLimitY = tan(0.25f * clampedLimitY);
+        const double tanQuarterSwingLimitZ = tan(0.25f * clampedLimitZ);
+
+        // violation term
+        double objectiveViolation = 0.0;
+        const size_t numPoses = m_localRotationSamples.size();
+        if (numPoses > 0)
+        {
+            for (const auto& localRotationSample : m_localRotationSamples)
+            {
+                objectiveViolation +=
+                    SwingValues(parentLocalConjugate * localRotationSample).GetViolation(tanQuarterSwingLimitY, tanQuarterSwingLimitZ);
+            }
+            objectiveViolation /= double(numPoses);
+        }
+
+        // volume term
+        const double objectiveVolume = 0.1 * swingLimitY * swingLimitZ;
```

Comment:
Nit: Might be worth some comments here stating the two terms and what they are and why they are there.

Reply:
Added a comment.